cloudimg/ms_azure.py

```diff
@@ -0,0 +1,480 @@
+from datetime import datetime, timedelta
+import logging
+import time
+
+import attr
+
+from azure.storage.blob import (
+    AccountSasPermissions,
+    BlobServiceClient,
+    BlobSasPermissions,
+    generate_account_sas,
+    generate_blob_sas,
+    ResourceTypes,
+)
+
+from cloudimg.common import BaseService, PublishingMetadata
+
+log = logging.getLogger(__name__)
+
+
+def get_values_from_connection_string(connection_str):
+    """
+    Get all the values from the connection string.
+
+    Args:
+        connection_string (str): The connection string of the
+                                 storage account
+
+    Returns:
+        A dictionary with all the key-values in the connection string
+    """
+    res = {}
+    val_list = connection_str.split(";")
+    for val in val_list:
+        if "=" in val:
+            entry = val.split("=", 1)
+            res[entry[0]] = entry[1]
+    return res
+
+
+def get_tags_filter_expression(tags):
+    """
+    Create a filter expression from the given tags for search.
+
+    The filter expression is a string with format:
+    ```
+    "tag1"='value1' and "tag2"='value2' ...
+    ```
+
+    Args:
+        tags (dict): Dicionary of strings with the tags to filter
+
+    Returns:
+        String with the filter expression.
+    """
+
+    # Create the filter expression like:
+    # "tag1"='value1" and "tag2" = 'value2" ...
+    filter_expression = ""
+    for k, v in tags.items():
+        filter_expression += "\"%s\"='%s' and " % (k, v)
+    # strip the last ' and '
+    return filter_expression.rstrip(" and ")
+
+
+@attr.s
+class UploadProgress(object):
+    """
+    Responsible for storing the upload progress.
+
+    It's expected to be instantiated by a callback function during blob
+    upload phase.
+
+    Args:
+        current (float): The current uploaded size in bytes
+        total (float): The total size to upload in bytes
+        percentage (float, optional): The upload progress percentage
+    """
+
+    current = attr.ib(type=float, repr=lambda x: "%.2fMB" % (x / 1024**2))
+    total = attr.ib(type=float, repr=lambda x: "%.2fMB" % (x / 1024**2))
+    percentage = attr.ib(type=float, repr=lambda x: "%.2f%%" % (x))
+    last_seen = attr.ib(type=float, repr=False, eq=False, factory=time.time)
```

Comment:
time.time isn't really the right way to do things like this because:

- it doesn't always move forwards
- even when moving forwards, it doesn't necessarily do so at the rate of one second per second

For measuring the time between two events, time.monotonic is more appropriate.

Reply:
Thanks a lot for this, I wasn't aware of it.

I've changed the factory to use `time.monotonic` but it seems like `py27` didn't like it.

Is it ok to leave like this as we will drop `py27` soon or should I try some workaround on it?

Reply:
Update: I found this [library](https://github.com/atdt/monotonic) which covers it.

I think it's better to have the pipeline passing, if no need please let me know then I remove it.